### ISSUE TITLE
Memory increases over time

### ISSUE THREAD

**Issue by djbokkeman** [RediSearch/RediSearch#874]
This is a follow-up on #866 - memory increasing in time.

Currently we see (virtual)memory increasing in time. It seems to be happening when calling:
1. Have a operational database with plenty of keys
2. Call: "FT.DROP indexname KEEPDOCS"
3. Call: "FT.CREATE indexname STOPWORDS 0 SCHEMA"
4. Call: "FT.ADDHASH indexname product:x 1 REPLACE LANGUAGE dutch" (x thousands of products)
5. Repeat step 2.

After dropping the index and re-adding all hashes the virtual memory increases in size. A restart of the Redis-server seems to solve this.

I got a .rdb from #866 available if anyone is interested.

**Comment by MeirShpilraien** [comment]
@djbokkeman did you eventually tested with this branch: https://github.com/RediSearch/RediSearch/pull/857?

**Comment by djbokkeman** [comment]
Hi @MeirShpilraien, Sorry my bad. I tested it on the 1.4 (thought it contained the fixes). 
Now I just compiled and loaded #857 and it seems to be improved a lot but it's still expanding memory. While the new index get's recreated we also regenerate all hashes and adding them to the index and also adding search suggestions etc. Everything we add is provided with the REPLACE argument.

Please let me know if I can help further debugging.

**Comment by MeirShpilraien** [comment]
@djbokkeman I will try to reproduce with the steps you describe under valgrind. Will update you if I will find something. If I will find nothing we will see how we continue with the investigation.

**Comment by MeirShpilraien** [comment]
@djbokkeman are you using different indexes names when re-creating the index? If so I found a big leak when dropping an index. Can you re-check with this branch: https://github.com/RediSearch/RediSearch/pull/857 (notice that I force pushed to it cause I rebased from 1.4).
Pls let me know how it looks.

**Comment by djbokkeman** [comment]
@MeirShpilraien yes we are using a lot of indexes. I will re-check the branch and let you know!

**Comment by djbokkeman** [comment]
Hey @MeirShpilraien it seems to help but it's stil expanding the memory in our use-case.

**Comment by djbokkeman** [comment]
@MeirShpilraien I wil happily provide you a new dump. Would It be helpful to provide an Appendonly database ?

**Comment by MeirShpilraien** [comment]
Yes AOF will be great, will help to reproduce the leak.

**Comment by MeirShpilraien** [comment]
@djbokkeman any news ?

**Comment by djbokkeman** [comment]
Hi @MeirShpilraien, I did not had time yet to set everything up. But I will try this or next week.

**Comment by MeirShpilraien** [comment]
@djbokkeman I have added another commit to this branch: https://github.com/RediSearch/RediSearch/pull/857

There was a memory leak caused by not calling pthread_detach on the GC and INDEXER threads. This means that those threads was stopped successfully but the memory they take was not free because it was kept in case someone will call join at some point. Calling pthread_detach fix it.

If you have time to check again please let me know if it helped..

**Comment by djbokkeman** [comment]
Hi @MeirShpilraien, sorry for my late response. I just updated to 1.4 and let you know how memory behaves.

**Comment by djbokkeman** [comment]
@MeirShpilraien It went up from 6gb to 14gb in a few hours. I'm a little bit afraid to run the database in AOF mode but will try.

**Comment by MeirShpilraien** [comment]
@djbokkeman 
Wowo, is it worst then before?

**Comment by djbokkeman** [comment]
Hi @MeirShpilraien, It's stable for a few hours now without restarts around 14gb. I also started saving the AOF. I will try to let it run till tomorrow without restarts and send you the dump+AOF. 

It seems to increase at a restart rapidly and around 13-14gb and then it stops expanding. Previously it would grow till 70gb and a restart was necessary. So it might just be allright now. Keep you posted!

**Comment by MeirShpilraien** [comment]
This is a good news, thanks you, let us know how it passed the night.

**Comment by djbokkeman** [comment]
@MeirShpilraien Live database grew from 14,7 to 15.9 gb overnight. It's a small increase compared to a few releases ago. We did activate AOF on our staging. But because it's our staging environment it takes a lot of time to increase in memory because of absence of traffic. And we have trouble to simulate this kind of traffic. I will pack the database and send you a link. After a few hours we could not restart the database (spawn error) that was saved as AOF. And we had 2000+ files beginning with temp-rewriteaof-xxxx.aof as filename.

**Comment by djbokkeman** [comment]
@MeirShpilraien just send you an invite for downloading the database. Thanks for the help again!

**Comment by MeirShpilraien** [comment]
@djbokkeman I will look at it.
